### PR TITLE
Be consistent with solidity merkle tree library used in plasma contracts

### PR DESCRIPTION
**Pull request by pgebal**
- Calculates merkle tree and proofs as in solidity library (solving issue https://github.com/omisego/plasma-contracts/issues/425)
- Adds `0` byte to each leaf before hashing and `1` byte to each node before hashing.
- Removes deprecated `proven?/3` function.

### CODE REVIEW

**Comment by boolafish** [comment]
Can we update the readme too? Or let's create issue to document our current Merkle Tree somewhere.

File: test/merkle_tree_test.exs [hunk 1]
```diff
@@ -27,50 +30,43 @@ defmodule MerkleTreeTest do
...
-    blocks = ['a', 'a', 'a', 'a']
+    blocks = ["a", "a", "a", "a"]
     assert MerkleTree.new(blocks, &MerkleTree.Crypto.sha256/1) == MerkleTree.new(blocks)
   end
 
   test "merkle tree node" do
     hash = &MerkleTree.Crypto.sha256/1
 
-    assert MerkleTree.build(['a', 'b'], &MerkleTree.Crypto.sha256/1) == %MerkleTree.Node{
+    assert MerkleTree.build(["a", "b"], &MerkleTree.Crypto.sha256/1) == %MerkleTree.Node{
```
Comment (pnowosie) [inline test/merkle_tree_test.exs]: please use the variable everywhere
```suggestion
    assert MerkleTree.build(["a", "b"], hash) == %MerkleTree.Node{
```

File: lib/merkle_tree.ex [hunk 2]
```diff
@@ -7,23 +7,57 @@ defmodule MerkleTree do
...
+        value: "9dc1674ae1ee61c90ba50b6261e8f9a47f7ea07d92612158edfe3c2a37c6d74c"
+      }
+    }
   """
 
   defstruct [:blocks, :root, :hash_function]
 
   # Number of children per node. Configurable.
   @number_of_children 2
+  @leaf_salt <<0>>
```
Comment (pnowosie) [inline lib/merkle_tree.ex:62]: I think it deserves a comment what are this values.

File: lib/merkle_tree.ex [hunk 3]
```diff
@@ -70,10 +103,10 @@ defmodule MerkleTree do
   """
   @spec fast_root(blocks, Keyword.t()) :: MerkleTree.Node.hash()
   def fast_root(blocks, opts \\ []) do
-    {hash_function, height, hash_leaves?, default_data_block} = get_from_options(opts, blocks)
+    {hash_function, height, default_data_block} = get_from_options(opts, blocks)
 
-    default_leaf_value = if hash_leaves?, do: hash_function.(default_data_block), else: default_data_block
```
Comment (pnowosie) [inline lib/merkle_tree.ex:75]: At first I thought that we should keep this option as we're planning to open :pr: in the source repo...
But at the 2nd thought I think adding this option was a bad idea at the beginning.

File: lib/merkle_tree.ex [hunk 4]
```diff
@@ -83,29 +116,28 @@ defmodule MerkleTree do
 
   defp _fast_root([root], _, final_height, _, final_height), do: root
 
-  defp _fast_root(nodes, hash_function, height, default_leaf, final_height) do
+  defp _fast_root(nodes, hash_function, height, default_node, final_height) do
```
Comment (pnowosie) [inline lib/merkle_tree.ex:122]: :+1:

**Review by pnowosie — APPROVED** [review]
Just minor suggestions

**Comment by pnowosie** [comment]
> Can we update the readme too?

We have plans to contribute this to the source repo  as well. So at some point it'll get better.
Overall quality of the code is let say `very hacky`

**Pull request merged**